File: 2015/python/d12.py
import json

def load(file):
    with open(file) as f:
        return json.load(f)

def solve(d, part2):
    sum = 0
    if type(d) is dict:
        for key in d.keys():
            sum += solve(d[key], part2)
    elif type(d) is list:
        if (not part2) or (not 'red' in d):
            for key in d:
                sum += solve(key, part2)
    else:
        try:
            i = int(d)
            sum += i
        except:
            i = 0
    return sum

def part1(d):
    return solve(d, False)

def part2(d):
    return solve(d, True)

d = load('../inputs/input_12.txt')
print(part1(d))
print(part2(d))

File: 2015/R/d12.R
d <- rjson::fromJSON(file = "../inputs/input_12.txt")

number <- function(x) {
  suppressWarnings(ifelse(!is.na(as.numeric(x)), as.numeric(x), 0))
}

count <- function(x, part2 = FALSE) {
  if (part2 & (length(names(x)) > 0)) {
    for (x2 in names(x)) {
      if (length(x[[x2]]) == 1) {
        if (x[[x2]] == "red") {
          return(0)
        }
      }
    }
  }
  ifelse(((class(x) == "list") || (length(x) > 1)),
         sum(unlist(lapply(x, count, part2))),
         number(x))
}
c(count(d), count(d, TRUE))
